### ISSUE TITLE
[Messenger]: Doctrine transport seems to be generating excessive number of queries

### ISSUE THREAD

**Issue by mareksuscak** [symfony/symfony#33659]
**Symfony version(s) affected**: 4.3.4

**Description**  

We've recently migrated away from [JMS Job Queue Bundle](https://packagist.org/packages/jms/job-queue-bundle) towards Messenger as part of our stack standardization effort. We set up 3 async priority transports (Doctrine type) and are running about 4 worker processes that all consume messages from all 3 transports. Immediately after initially deploying it to staging, we started seeing huge spike in the number of MySQL queries (we're using Google Cloud SQL v5.7). 

**How to reproduce**  

1. Create a new Symfony app
2. Install messenger and configure it with 3 doctrine based transports according to the config file attached below
3. Run the command attached below
4. Check the QPS

**Possible Solution**  

N/A

**Additional context**  

```sh
# shell command
$ bin/console messenger:consume async_priority_high async async_priority_low --time-limit=86400 --memory-limit=128M -vv
```

```
# .env
MESSENGER_TRANSPORT_DSN=doctrine://default
```

```yaml
# config/packages/messenger.yaml
framework:
    messenger:
        # Send failed messages to this transport for later handling.
        failure_transport: failed

        transports:
            # https://symfony.com/doc/current/messenger.html#transport-configuration

            # Use for high priority tasks such as search jobs
            async_priority_high:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                retry_strategy:
                    max_retries: 3
                    # milliseconds delay
                    delay: 1000
                    # causes the delay to be higher before each retry
                    # e.g. 1 second delay, 2 seconds, 4 seconds
                    multiplier: 2
                    max_delay: 0
                options:
                    queue_name: high_priority
            # Use this as a default or when you're not sure
            async:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    queue_name: normal_priority
            # Use for low priority tasks such as sending emails, indexing...
            async_priority_low:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    queue_name: low_priority
            # This is the default
            sync: 'sync://'
            # All failed messages end up here and can be retried
            failed:
                dsn: '%env(MESSENGER_TRANSPORT_DSN)%'
                options:
                    queue_name: failed

        routing:
            # Route your messages to the transports
            'App\Message\ServiceSearchRequest': async_priority_high
            'App\Message\HotelSearchRequest': async_priority_high
            'App\Message\ZumataHotelProcessRequest': async_priority_high
            'App\Message\HotelSearchFinishRequest': async_priority_low
            # Mailer messages
            'Symfony\Component\Mailer\Messenger\SendEmailMessage': async

        # Read more to learn why there are two buses (TL;DR: different middleware on each side - producer and consumer)
        # https://github.com/symfony/symfony/issues/32436#issuecomment-530565499

        default_bus: command.bus
        buses:
            command.bus:
                default_middleware: false
                middleware:
                    - validation
                    - add_bus_name_stamp_middleware: ['command.consumer.bus']
                    - dispatch_after_current_bus
                    - failed_message_processing_middleware
                    - send_message
                    - handle_message

            command.consumer.bus:
                middleware:
                    - doctrine_ping_connection
                    # Will be available as of Symfony 4.4
                    # - doctrine_clear_entity_manager
                    # TODO: we need this middleware for all but HotelSearchRequest and ServiceSearchRequest
                    # we can create two different buses and use a transaction-free one explicitly for these
                    # two cases
                    # - doctrine_transaction

```

Here's a chart that shows no. of MySQL queries per second. All 4 workers running in a staging environment, no messages are in the queue:

![image](https://user-images.githubusercontent.com/613647/65374569-b662bd80-dc8b-11e9-999f-004c6c66e9af.png)

Here's a chart that shows no. of MySQL queries per second. Only 1 worker running in a staging environment, no messages are in the queue:

![image](https://user-images.githubusercontent.com/613647/65374590-ed38d380-dc8b-11e9-85eb-256c1634e4a8.png)

**Comment by alister** [comment]
Looks to be because of https://github.com/symfony/symfony/pull/32308

**Comment by mareksuscak** [comment]
@alister tried it, didn't help at all. Here's the chart again, 3 workers using 3 transports each.

<img width="1398" alt="Screen Shot 2019-09-23 at 10 42 51" src="https://user-images.githubusercontent.com/613647/65412213-0e660500-ddef-11e9-9291-ece837b2c01c.png">

**Comment by mareksuscak** [comment]
So I added two DB flags - `general_log: On` and `log_output: FILE` and it seems that some meta queries are being generated times and times again. There's a recurring pattern and here are some of them:

```sql
# Query 1
SELECT ENGINE, AUTO_INCREMENT, TABLE_COLLATION, TABLE_COMMENT, CREATE_OPTIONS FROM information_schema.TABLES WHERE TABLE_TYPE = 'BASE TABLE' AND TABLE_SCHEMA = DATABASE() AND TABLE_NAME = 'exchange_rate';

# Query 2
SELECT COLUMN_NAME AS Field, COLUMN_TYPE AS Type, IS_NULLABLE AS `Null`, COLUMN_KEY AS `Key`, COLUMN_DEFAULT AS `Default`, EXTRA AS Extra, COLUMN_COMMENT AS Comment, CHARACTER_SET_NAME AS CharacterSet, COLLATION_NAME AS Collation FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = 'spalneo_staging' AND TABLE_NAME = 'exchange_rate' ORDER BY ORDINAL_POSITION ASC;

# Query 3
SELECT DISTINCT k.`CONSTRAINT_NAME`, k.`COLUMN_NAME`, k.`REFERENCED_TABLE_NAME`, k.`REFERENCED_COLUMN_NAME` /*!50116 , c.update_rule, c.delete_rule */ FROM information_schema.key_column_usage k /*!50116 INNER JOIN information_schema.referential_constraints c ON c.constraint_name = k.constraint_name AND c.table_name = 'exchange_rate' */ WHERE k.table_name = 'exchange_rate' AND k.table_schema = 'spalneo_staging' /*!50116 AND c.constraint_schema = 'spalneo_staging' */ AND k.`REFERENCED_COLUMN_NAME` is not NULL;

# Query 4
Query SELECT TABLE_NAME AS `Table`, NON_UNIQUE AS Non_Unique, INDEX_NAME AS Key_name, SEQ_IN_INDEX AS Seq_in_index, COLUMN_NAME AS Column_Name, COLLATION AS Collation, CARDINALITY AS Cardinality, SUB_PART AS Sub_Part, PACKED AS Packed, NULLABLE AS `Null`, INDEX_TYPE AS Index_Type, COMMENT AS Comment FROM information_schema.STATISTICS WHERE TABLE_NAME = 'exchange_rate' AND TABLE_SCHEMA = 'spalneo_staging' ORDER BY SEQ_IN_INDEX ASC;

# Query 5
SELECT TABLE_NAME AS `Table`, NON_UNIQUE AS Non_Unique, INDEX_NAME AS Key_name, SEQ_IN_INDEX AS Seq_in_index, COLUMN_NAME AS Column_Name, COLLATION AS Collation, CARDINALITY AS Cardinality, SUB_PART AS Sub_Part, PACKED AS Packed, NULLABLE AS `Null`, INDEX_TYPE AS Index_Type, COMMENT AS Comment FROM information_schema.STATISTICS WHERE TABLE_NAME = 'ext_translations' AND TABLE_SCHEMA = 'spalneo_staging' ORDER BY SEQ_IN_INDEX ASC;
```

**Comment by Tobion** [comment]
Messenger does not do those meta queries. So it must come from somewhere else.

**Comment by mareksuscak** [comment]
They only appear in the logs while workers are running. When I stop them, there are no such queries. So there must be something that the workers trigger.

**Comment by mareksuscak** [comment]
Alright so after further investigation it appears that these queries stem from `doctrine/dbal`.

- Query 1
 https://github.com/doctrine/dbal/blob/b0726e7aab01080fc385599a88491a0ccdbd9417/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L383
- Query 2 https://github.com/doctrine/dbal/blob/b0726e7aab01080fc385599a88491a0ccdbd9417/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L366
- Query 3 https://github.com/doctrine/dbal/blob/b0726e7aab01080fc385599a88491a0ccdbd9417/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L180
- Query 4 https://github.com/doctrine/dbal/blob/22800bd651c1d8d2a9719e2a3dc46d5108ebfcc9/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L151

**Comment by alister** [comment]
Yes, as the first message in the [other thread](https://github.com/symfony/symfony/pull/32308#issue-293352651) says '... because a schema diff is done in setup every get.'   At the end of all of the the `SELECT TABLE_NAME AS ...`, it then deletes and re-creates an index on the table. `CREATE INDEX IDX_random_name ON messenger_messages (queue_name)`. 

Only after this does it do the expected search:

```
START TRANSACTION
SELECT m.* FROM messenger_messages m WHERE ...
COMMIT
```

When auto_setup is turned off for *all* queues, then it will only do the search & any updates (in a transaction) for the actual messenger.  I can turn the schema-diffing (which is what takes all the effort) on and off at will, by setting `%env(string:MESSENGER_TRANSPORT_DSN)%` to `doctrine://default?auto_setup=false`, and having that file auto-reloaded by my local setup.

Use `bin/console debug:container  --env-var=MESSENGER_TRANSPORT_DSN` to show the value of the variable from the command line, but also try to also show the value from within the program.  If you have a value like `doctrine://default?auto_setup=false` in one place, but not inside the script, it might not have been properly updated in the local environment.  (that can happen in a a local dev setup, if you edit the .env file, but don't re-run it in your shell to being the latest versions into the environment - because the dotenv system will prefer the environment variables over the file).

**Comment by Tobion** [comment]
Closing as duplicate of #32308

**Comment by mareksuscak** [comment]
> Use `bin/console debug:container --env-var=MESSENGER_TRANSPORT_DSN` to show the value of the variable from the command line, but also try to also show the value from within the program. If you have a value like `doctrine://default?auto_setup=false` in one place, but not inside the script, it might not have been properly updated in the local environment. (that can happen in a a local dev setup, if you edit the .env file, but don't re-run it in your shell to being the latest versions into the environment - because the dotenv system will prefer the environment variables over the file).

Shoot. I should've done this before. Your comment made me realize that we're injecting real env vars into the staging and prod environments since the app is running in a docker container and these env vars obviously override the contents of `.env` file. And guess who set it up… 🙌🙈 This is somewhat embarrassing to admit. Thanks for pointing it out! 

**Staging:**
![image](https://user-images.githubusercontent.com/613647/65464548-5bcb8c00-de5a-11e9-8b8f-282f0b4bf053.png)

**Prod:**
![image](https://user-images.githubusercontent.com/613647/65464704-c41a6d80-de5a-11e9-8c85-6f8579ffb5fb.png)
